[ros2node] info fails to find nodes in namespaces

<!--
For general questions, please ask on ROS answers: https://answers.ros.org, make sure to include at least the `ros2` tag and the rosdistro version you are running, e.g. `ardent`.
For general design discussions, please post on discourse: https://discourse.ros.org/c/ng-ros
Not sure if this is the right repository? Open an issue on https://github.com/ros2/ros2/issues
For Bug report or feature requests, please fill out the relevant category below
-->

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 18.04
- Installation type:
  - Source
- Version or commit hash:
  - ros2cli 7a65e1ecf02c69bf3d9624d49b17cde57f7ed16a
  - rmw_fastrtps d8f1959a785ad770f96623d257b8a4ae1c74c938
- DDS implementation:
  - Fast-RTPS
- Client library (if applicable):
  - N/A

#### Steps to reproduce issue
<!-- Detailed instructions on how to reliably reproduce this issue http://sscce.org/
``` code that can be copy-pasted is preferred ``` -->

1. Save script as `namespace.launch.py`
```
#!/usr/bin/env python3

from launch import LaunchDescription

from launch_ros.actions import Node


def generate_launch_description():
    talker1 = Node(package='demo_nodes_cpp', node_executable='talker')
    talker2 = Node(package='demo_nodes_cpp', node_executable='talker', node_namespace='ns')

    return LaunchDescription([talker1, talker2])
```

2. Run `ros2 launch namespace.launch.py`

3. Run `ros2 node info /talker`

4. Run `ros2 node info /ns/talker`

#### Expected behavior
Information is printed for both nodes.

#### Actual behavior
Node inside a namespace throws
```
[ERROR] [rmw_fastrtps_shared_cpp]: Unable to find GUID for node: talker
Failed to get_subscriber_names_and_types: Unable to find GUID for node , at /home/jacob/ros2_core_ws/src/ros2/rmw_fastrtps/rmw_fastrtps_shared_cpp/src/rmw_node_info_and_types.cpp:85
```

#### Additional information

--- a/ros2node/ros2node/api/__init__.py
+++ b/ros2node/ros2node/api/__init__.py
@@ -29,16 +29,18 @@ def get_absolute_node_name(node_name):
     return node_name
 
 
-def parse_node_name(full_node_name):
-    tokens = full_node_name.split('/')
-    if 1 > len(tokens):
-        raise RuntimeError('Invalid node name: ' + full_node_name)
-    node_name = full_node_name
-    namespace = '/'
+def parse_node_name(node_name):
+    if (node_name[0] == '/'):
+        node_name = node_name[1:]
+    tokens = node_name.split('/')
+    node_basename = tokens[-1]
     if len(tokens) > 1:
-        node_name = tokens[-1]
         namespace = '/' + '/'.join(tokens[:-1])
-    return NodeName(node_name, namespace, '/' + full_node_name)
+        full_node_name = namespace + '/' + node_basename
+    else:
+        namespace = '/'
+        full_node_name = '/' + node_basename
+    return NodeName(node_basename, namespace, full_node_name)
 
 
 def get_node_names(*, node, include_hidden_nodes=False):

--- a/ros2node/ros2node/api/__init__.py
+++ b/ros2node/ros2node/api/__init__.py
@@ -30,16 +30,15 @@ def get_absolute_node_name(node_name):
 
 
 def parse_node_name(node_name):
-    if (node_name[0] == '/'):
-        node_name = node_name[1:]
-    tokens = node_name.split('/')
-    node_basename = tokens[-1]
-    if len(tokens) > 1:
-        namespace = '/' + '/'.join(tokens[:-1])
-        full_node_name = namespace + '/' + node_basename
+    if node_name[0] != '/':
+        full_node_name = '/' + node_name
     else:
+        full_node_name = node_name
+    tokens = full_node_name.rsplit('/', 1)
+    node_basename = tokens[1]
+    namespace = tokens[0]
+    if namespace == '':
         namespace = '/'
-        full_node_name = '/' + node_basename
     return NodeName(node_basename, namespace, full_node_name)
 
 

--- a/ros2node/ros2node/api/__init__.py
+++ b/ros2node/ros2node/api/__init__.py
@@ -37,8 +37,8 @@ def parse_node_name(full_node_name):
     namespace = '/'
     if len(tokens) > 1:
         node_name = tokens[-1]
-        namespace = '/'.join(tokens[:-1])
-    return NodeName(node_name, namespace, full_node_name)
+        namespace = '/' + '/'.join(tokens[:-1])
+    return NodeName(node_name, namespace, '/' + full_node_name)
 
 
 def get_node_names(*, node, include_hidden_nodes=False):

--- a/ros2node/ros2node/api/__init__.py
+++ b/ros2node/ros2node/api/__init__.py
@@ -30,13 +30,10 @@ def get_absolute_node_name(node_name):
 
 
 def parse_node_name(node_name):
-    if node_name[0] != '/':
-        full_node_name = '/' + node_name
-    else:
-        full_node_name = node_name
-    tokens = full_node_name.rsplit('/', 1)
-    node_basename = tokens[1]
-    namespace = tokens[0]
+    full_node_name = node_name
+    if not full_node_name.startswith('/'):
+        full_node_name = '/' + full_node_name
+    namespace, node_basename = full_node_name.rsplit('/', 1)
     if namespace == '':
         namespace = '/'
     return NodeName(node_basename, namespace, full_node_name)

--- a/ros2node/ros2node/verb/info.py
+++ b/ros2node/ros2node/verb/info.py
@@ -43,13 +43,13 @@ class InfoVerb(VerbExtension):
         if args.node_name in [n.full_name for n in node_names]:
             with DirectNode(args) as node:
                 print(args.node_name)
-                subscribers = get_subscriber_info(node=node, remote_node_name=args.node_name[1:])
+                subscribers = get_subscriber_info(node=node, remote_node_name=args.node_name)
                 print('  Subscribers:')
                 print_names_and_types(subscribers)
-                publishers = get_publisher_info(node=node, remote_node_name=args.node_name[1:])
+                publishers = get_publisher_info(node=node, remote_node_name=args.node_name)
                 print('  Publishers:')
                 print_names_and_types(publishers)
-                services = get_service_info(node=node, remote_node_name=args.node_name[1:])
+                services = get_service_info(node=node, remote_node_name=args.node_name)
                 print('  Services:')
                 print_names_and_types(services)
         else:

--- a/ros2node/test/test_node.py
+++ b/ros2node/test/test_node.py
@@ -18,6 +18,14 @@ from ros2node.api import parse_node_name
 def test_parse_node_name():
     """Test parse_node_name api function."""
     # The input of parse_node_name is the full_name without the initial '/'
+    name = parse_node_name('/talker')
+    assert name.full_name == '/talker'
+    assert name.namespace == '/'
+    assert name.name == 'talker'
+    name = parse_node_name('/ns/talker')
+    assert name.full_name == '/ns/talker'
+    assert name.namespace == '/ns'
+    assert name.name == 'talker'
     name = parse_node_name('talker')
     assert name.full_name == '/talker'
     assert name.namespace == '/'

--- a/ros2node/test/test_node.py
+++ b/ros2node/test/test_node.py
@@ -1,0 +1,28 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ros2node.api import parse_node_name
+
+
+def test_parse_node_name():
+    """Test parse_node_name api function."""
+    # The input of parse_node_name is the full_name without the initial '/'
+    name = parse_node_name('talker')
+    assert name.full_name == '/talker'
+    assert name.namespace == '/'
+    assert name.name == 'talker'
+    name = parse_node_name('ns/talker')
+    assert name.full_name == '/ns/talker'
+    assert name.namespace == '/ns'
+    assert name.name == 'talker'

